src/geometry.rs

```diff
@@ -244,6 +250,138 @@ impl Geometry {
         }
         used_normals.len() == self.normals().len()
     }
+
+    /// Calculates topological relations (neighborhood) of mesh face -> faces.
+    /// Returns a Map (key: face index, value: list of its neighboring faces indices)
+    pub fn face_to_face_topology(&self) -> HashMap<usize, SmallVec<[usize; 16]>> {
+        let mut f2f: HashMap<usize, SmallVec<[usize; 16]>> = HashMap::new();
+        for (from_counter, f) in self.triangle_faces_iter().enumerate() {
+            let [f_e_0, f_e_1, f_e_2] = f.to_unoriented_edges();
+            for (to_counter, t_f) in self.triangle_faces_iter().enumerate() {
+                if from_counter != to_counter && t_f.contains_unoriented_edge(f_e_0)
+                    || t_f.contains_unoriented_edge(f_e_1)
+                    || t_f.contains_unoriented_edge(f_e_2)
+                {
+                    let neighbors = f2f.entry(from_counter).or_insert_with(SmallVec::new);
+                    if neighbors.iter().all(|value| *value != to_counter) {
+                        neighbors.push(to_counter);
+                    }
+                }
+            }
+        }
+        f2f
+    }
+
+    /// Calculates topological relations (neighborhood) of mesh edge -> faces.
+    /// Returns a Map (key: edge index, value: list of its neighboring faces indices)
+    pub fn edge_to_face_topology(
+        &self,
+        edges: &HashSet<UnorientedEdge>,
+    ) -> HashMap<usize, SmallVec<[usize; 16]>> {
+        let mut e2f: HashMap<usize, SmallVec<[usize; 16]>> = HashMap::new();
+        for (from_counter, e) in edges.iter().enumerate() {
```

Comment:
This is dubious. `HashSet`s have randomization and reseeding, so no two instances will contain the elements in the same order.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=use%20std%3A%3Acollections%3A%3AHashSet%3B%0Ause%20std%3A%3Aiter%3A%3AFromIterator%3B%0A%0Afn%20main()%20%7B%0A%20%20%20%20let%20data%3A%20Vec%3Cu8%3E%20%3D%20vec!%5B0%2C%201%2C%202%2C%203%2C%204%2C%205%2C%206%2C%207%2C%208%2C%209%5D%3B%0A%20%20%20%20let%20set1%3A%20HashSet%3Cu8%3E%20%3D%20HashSet%3A%3Afrom_iter(data.iter().copied())%3B%0A%20%20%20%20let%20set2%3A%20HashSet%3Cu8%3E%20%3D%20HashSet%3A%3Afrom_iter(data.iter().copied())%3B%0A%20%20%20%20%0A%20%20%20%20println!(%22--%20SET1%20--%22)%3B%0A%20%20%20%20for%20(i%2C%20elem)%20in%20set1.iter().enumerate()%20%7B%0A%20%20%20%20%20%20%20%20println!(%22%7B%7D%3A%20%7B%7D%22%2C%20i%2C%20elem)%3B%0A%20%20%20%20%7D%0A%0A%20%20%20%20println!(%22--%20SET2%20--%22)%3B%0A%20%20%20%20for%20(i%2C%20elem)%20in%20set2.iter().enumerate()%20%7B%0A%20%20%20%20%20%20%20%20println!(%22%7B%7D%3A%20%7B%7D%22%2C%20i%2C%20elem)%3B%0A%20%20%20%20%7D%0A%7D

Will we be able to guarantee we always have the same instance of the set around? Seems tricky to do. Shouldn't we have a way of creating and using the edges in a stable order? I imagine they could be a `Vec`, but we would use a `HashSet` during the creation to perform deduplication.

Also applies to the other method using edges.

Reply:
this can be any `Iterator` or directly a `Vec` in fact. it will work for duplicates as well, so it can be the responsibility of the caller to purge or keep the duplicates.